### PR TITLE
feat(auth): Add auth emulator support

### PR DESCRIPTION
**Pull request by ssbushi**
### Discussion

Add support to communicate with the Auth Emulator #493.

### Testing

  * Unit tests.
  * Manual testing.

### API Changes

  * N/A.

### CODE REVIEW

File: src/test/java/com/google/firebase/auth/multitenancy/FirebaseTenantClientTest.java [hunk 1]
```diff
@@ -312,6 +321,20 @@ public void testDeleteTenantWithNotFoundError() {
     checkUrl(interceptor, "DELETE", TENANTS_BASE_URL + "/UNKNOWN");
   }
 
+  @Test
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/auth/multitenancy/FirebaseTenantClientTest.java:324]: Ideally we need a way to test all APIs in the class with the emulator.

File: src/test/java/com/google/firebase/auth/FirebaseTokenVerifierImplTest.java [hunk 2]
```diff
@@ -399,20 +399,6 @@ public void testVerifyTokenMissingTenantId() {
     }
   }
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/auth/FirebaseTokenVerifierImplTest.java:401]: Why was this removed?

File: src/test/java/com/google/firebase/auth/FirebaseEmulatorTokenVerifierTest.java [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.firebase.ErrorCode;
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import org.junit.Before;
+import org.junit.Test;
+
+public class FirebaseEmulatorTokenVerifierTest {
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/auth/FirebaseEmulatorTokenVerifierTest.java:42]: Most of this can be combined with the existing tests once the emulator support is added to the existing `FirebaseTokenVerfierImpl` class.

**Review by hiranya911 — CHANGES_REQUESTED** [review]
Thanks @ssbushi. Looks mostly good. My main observations are:

1. We should avoid duplication by adding emulator support to the existing `FirebaseTokenVerfierImpl` class, as opposed to implementing a whole new class.
2. Avoid duplication by putting env variable lookups in a shared util.
3. We need a better way to test the functionality of all APIs in emulator mode. In other languages we used techniques like test parameterization and fixtures to achieve this. We should try to find an equivalent strategy for junit4 (although it's ok to do so in a separate PR).

File: src/test/java/com/google/firebase/auth/FirebaseTokenVerifierImplTest.java (hunk 2 above)
Comment (ssbushi) [reply src/test/java/com/google/firebase/auth/FirebaseTokenVerifierImplTest.java:401]: Sorry, I forgot to add a comment here.

This test passes, but only because the `verifyToken` does not throw. The current behaviour is that the token verifier silently ignores the `tenantId` in the input token, if the verfier itself is not built with a `tenantId` set.

I was wondering if this is the intended behavior or if we expect the verifier to throw (in which case I can update the implementation and add a `fail()` in the test's try block). 

I haven't tested other tests in this file, but since they lack a `fail()` in the try block, I suspect that there might be a few that are passing regardless of the test conditions.

Comment (hiranya911) [reply src/test/java/com/google/firebase/auth/FirebaseTokenVerifierImplTest.java:401]: The verifier is initialized with a tenantId in this case if I'm not mistaken:

```java
fullyPopulatedBuilder()
        .setTenantId("TENANT_ID") // <----
        .build()
        .verifyToken(tokenFactory.createToken());
```

We should be able to add `fail()` statement to the `try` block.

Comment (ssbushi) [reply src/test/java/com/google/firebase/auth/FirebaseTokenVerifierImplTest.java:401]: I'm sorry, I don't get it. This test here does not have a verifier with a tenantId.

```
fullyPopulatedBuilder()
        .build()
        .verifyToken(createTokenWithTenantId("TENANT_ID"));
```


This test is same as the `testVerifyTokenWithTenantId` but it's expecting a different behaviour.

Comment (hiranya911) [reply src/test/java/com/google/firebase/auth/FirebaseTokenVerifierImplTest.java:401]: This is testing a regular token against a tenant-aware verifier (this should throw):

```
fullyPopulatedBuilder()
        .setTenantId("TENANT_ID")
        .build()
        .verifyToken(tokenFactory.createToken());
```

This is testing a tenant-scoped token against a tenant-unaware verifier:

```
fullyPopulatedBuilder()
        .build()
        .verifyToken(createTokenWithTenantId("TENANT_ID"));
```

This used to throw the same error, but that behavior was changed in #475. This test case seems like something left over from before the change. We can update it to reflect the current behavior.

File: src/test/java/com/google/firebase/auth/FirebaseTokenVerifierImplTest.java [hunk 4]
```diff
@@ -384,21 +425,6 @@ public void testVerifyTokenNoTenantId() {
...
-        .build()
-        .verifyToken(tokenFactory.createToken());
-    } catch (FirebaseAuthException e) {
-      assertEquals(AuthErrorCode.TENANT_ID_MISMATCH, e.getAuthErrorCode());
-      assertEquals(
-          "The tenant ID ('') of the token did not match the expected value ('TENANT_ID')",
-          e.getMessage());
-    }
-  }
-
```
Comment (ssbushi) [inline src/test/java/com/google/firebase/auth/FirebaseTokenVerifierImplTest.java:401]: Removing this test as this is the same as `testVerifyTokenNoTenantId` above.

File: src/main/java/com/google/firebase/auth/FirebaseEmulatorTokenVerifier.java [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+import com.google.common.base.Strings;
+import com.google.firebase.ErrorCode;
+import com.google.firebase.internal.Nullable;
+import java.io.IOException;
+
+/**
+ * The implementation of the {@link FirebaseTokenVerifier} interface, for the Auth Emulator. Can be
+ * customized to verify both Firebase ID tokens and session cookies.
+ */
+final class FirebaseEmulatorTokenVerifier implements FirebaseTokenVerifier {
```
[the remark this answers is not in the document]
Comment (ssbushi) [reply src/main/java/com/google/firebase/auth/FirebaseEmulatorTokenVerifier.java]: Thanks Hiranya.

I've followed your advice and moved the logic over to `FirebaseTokenVerifierImpl` although I didn't add `isEmulatorMode` as a member. Instead the emulator mode is activated by checking for the respective environment variable. I thought this is more seamless, however I'm curious what you think.

File: src/main/java/com/google/firebase/auth/FirebaseTokenUtils.java [hunk 6]
```diff
@@ -62,7 +63,8 @@ static FirebaseTokenFactory createTokenFactory(
       return new FirebaseTokenFactory(
           firebaseApp.getOptions().getJsonFactory(),
           clock,
-          CryptoSigners.getCryptoSigner(firebaseApp),
+          useEmulator() ? new EmulatorCryptoSigner()
```
[the remark this answers is not in the document]
Comment (ssbushi) [reply src/main/java/com/google/firebase/auth/FirebaseTokenUtils.java]: Done.

File: src/main/java/com/google/firebase/auth/FirebaseUserManager.java [hunk 7]
```diff
@@ -100,6 +102,18 @@ private FirebaseUserManager(Builder builder) {
...
+  private String getIdToolkitUrl(String projectId, String version) {
+    String maybeEmulator = getEmulatorHost();
+    if (maybeEmulator != null) {
+      return String.format(ID_TOOLKIT_URL_EMULATOR, maybeEmulator, version, projectId);
+    }
+    return String.format(ID_TOOLKIT_URL, version, projectId);
+  }
+
+  private String getEmulatorHost() {
+    return System.getenv("FIREBASE_AUTH_EMULATOR_HOST");
```
[the remark this answers is not in the document]
Comment (ssbushi) [reply src/main/java/com/google/firebase/auth/FirebaseUserManager.java]: Done.

File: src/main/java/com/google/firebase/auth/FirebaseUserManager.java [hunk 8]
```diff
@@ -100,6 +102,18 @@ private FirebaseUserManager(Builder builder) {
     this.httpClient = new AuthHttpClient(jsonFactory, builder.requestFactory);
   }
 
+  private String getIdToolkitUrl(String projectId, String version) {
+    String maybeEmulator = getEmulatorHost();
+    if (maybeEmulator != null) {
```
[the remark this answers is not in the document]
Comment (ssbushi) [reply src/main/java/com/google/firebase/auth/FirebaseUserManager.java]: Done.

File: src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java [hunk 9]
```diff
@@ -2678,6 +2684,33 @@ public void testTenantAwareDeleteSamlProviderConfig() throws Exception {
     checkUrl(interceptor, "DELETE", expectedUrl);
   }
 
+  @Test
```
[the remark this answers is not in the document]
Comment (ssbushi) [reply src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java:2687]: Agreed. I propose to defer that to a new PR, to avoid making the current one too big.

File: src/test/java/com/google/firebase/auth/multitenancy/FirebaseTenantClientTest.java (hunk 1 above)
Comment (ssbushi) [reply src/test/java/com/google/firebase/auth/multitenancy/FirebaseTenantClientTest.java:324]: Agreed. See the previous comment.

File: src/test/java/com/google/firebase/auth/FirebaseEmulatorTokenVerifierTest.java (hunk 3 above)
Comment (ssbushi) [reply src/test/java/com/google/firebase/auth/FirebaseEmulatorTokenVerifierTest.java]: Done.

File: src/main/java/com/google/firebase/auth/FirebaseTokenUtils.java [hunk 10]
```diff
@@ -87,8 +87,9 @@ static FirebaseTokenVerifierImpl createIdTokenVerifier(
         clock, ID_TOKEN_ISSUER_PREFIX, projectId);
     GooglePublicKeysManager publicKeysManager = newPublicKeysManager(
         app.getOptions(), clock, ID_TOKEN_CERT_URL);
+    String shortName = Utils.isEmulatorMode() ? "Emulator ID token" : "ID token";
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/FirebaseTokenUtils.java:90]: I don't think this is necessary. Let's use the same error messages/text as the production mode. That's how this is implemented in other languages too.

File: src/main/java/com/google/firebase/auth/FirebaseTokenVerifierImpl.java [hunk 11]
```diff
@@ -96,7 +97,9 @@ private FirebaseTokenVerifierImpl(Builder builder) {
   public FirebaseToken verifyToken(String token) throws FirebaseAuthException {
     IdToken idToken = parse(token);
     checkContents(idToken);
-    checkSignature(idToken);
+    if (!isEmulatorMode()) {
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/FirebaseTokenVerifierImpl.java:100]: Personally I prefer the ability to programmatically enable the emulator mode, instead of just tying it to an environment variable. That makes unit testing a lot easier. However, this is the approach we've taken in our Node.js SDK so we can live with this.

But let's make sure we only lookup the env variable once per token verification. So assign the result to a local variable, and use it in rest of the method body:

```
boolean emulator = Utils.isEmulatorMode();
```

File: src/main/java/com/google/firebase/auth/Utils.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
+package com.google.firebase.auth;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+
+public class Utils {
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/Utils.java:6]: This class has to package-protected. If it must be public, then it should be moved to the `auth.internal` package.

File: src/main/java/com/google/firebase/auth/Utils.java [hunk 13]
```diff
@@ -0,0 +1,19 @@
+package com.google.firebase.auth;
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/Utils.java:1]: Add license header.

File: src/main/java/com/google/firebase/auth/Utils.java [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+
+public class Utils {
+  @VisibleForTesting
+  public static final String AUTH_EMULATOR_HOST = "FIREBASE_AUTH_EMULATOR_HOST";
+
+  public static boolean isEmulatorMode() {
+    return !Strings.isNullOrEmpty(
+        System.getenv(AUTH_EMULATOR_HOST));
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/Utils.java:12]: getEmulatorHost()

File: src/test/java/com/google/firebase/auth/TestTokenFactory.java [hunk 15]
```diff
@@ -86,4 +112,11 @@ public String createToken(JsonWebSignature.Header header, JsonWebToken.Payload p
     payload.setSubject(UID);
     return payload;
   }
+
+  public JsonWebSignature.Header createHeaderForEmulator() {
+    JsonWebSignature.Header header = new JsonWebSignature.Header();
+    header.setAlgorithm("NONE");
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/auth/TestTokenFactory.java:118]: It should be "none" as per the JWS spec.

File: src/test/java/com/google/firebase/auth/TestTokenFactory.java [hunk 16]
```diff
@@ -69,6 +70,31 @@ public String createToken(JsonWebSignature.Header header, JsonWebToken.Payload p
...
+  }
+
+  public String createUnsignedTokenForEmulator(JsonWebToken.Payload payload) {
+    return createUnsignedTokenForEmulator(createHeaderForEmulator(), payload);
+  }
+
+  public String createUnsignedTokenForEmulator(
+      JsonWebSignature.Header header, JsonWebSignature.Payload payload) {
+    try {
+      String content =
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/auth/TestTokenFactory.java:88]: Readability nit:

```
String encodedHeader = Base64.encodeBase64URLSafeString(JSON_FACTORY.toByteArray(header));
String encodedPayload = Base64.encodeBase64URLSafeString(JSON_FACTORY.toByteArray(payload));
return String.format("%s.%s.", encodedHeader, encodedPayload);
```

File: src/test/java/com/google/firebase/auth/FirebaseTokenVerifierImplTest.java [hunk 17]
```diff
@@ -443,6 +455,258 @@ public void testBuilderNoDocUrl() {
     fullyPopulatedBuilder().setDocUrl(null).build();
   }
 
+  @Test
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/auth/FirebaseTokenVerifierImplTest.java:458]: Might be just easier to create a new test class `EmulatorFirebaseTokenVerifierImpleTest` for all these new test cases. You can do the env set up in the `@Before` fixture.

**Review by hiranya911 — CHANGES_REQUESTED** [review]
Thanks for making the changes. Looks pretty good. Just a few more minor updates needed.

File: src/main/java/com/google/firebase/auth/FirebaseTokenUtils.java (hunk 10 above)
Comment (ssbushi) [reply src/main/java/com/google/firebase/auth/FirebaseTokenUtils.java]: Done.

File: src/main/java/com/google/firebase/auth/FirebaseTokenVerifierImpl.java (hunk 11 above)
Comment (ssbushi) [reply src/main/java/com/google/firebase/auth/FirebaseTokenVerifierImpl.java]: Done.

File: src/main/java/com/google/firebase/auth/Utils.java (hunk 13 above)
Comment (ssbushi) [reply src/main/java/com/google/firebase/auth/Utils.java]: Done.

File: src/main/java/com/google/firebase/auth/Utils.java (hunk 14 above)
Comment (ssbushi) [reply src/main/java/com/google/firebase/auth/Utils.java]: Done.

File: src/test/java/com/google/firebase/auth/TestTokenFactory.java (hunk 15 above)
Comment (ssbushi) [reply src/test/java/com/google/firebase/auth/TestTokenFactory.java]: Done.

File: src/test/java/com/google/firebase/auth/TestTokenFactory.java (hunk 16 above)
Comment (ssbushi) [reply src/test/java/com/google/firebase/auth/TestTokenFactory.java]: Done.

File: src/test/java/com/google/firebase/auth/FirebaseTokenVerifierImplTest.java (hunk 17 above)
Comment (ssbushi) [reply src/test/java/com/google/firebase/auth/FirebaseTokenVerifierImplTest.java]: Done.

File: src/main/java/com/google/firebase/auth/Utils.java (hunk 12 above)
Comment (ssbushi) [reply src/main/java/com/google/firebase/auth/Utils.java]: Done.

File: src/main/java/com/google/firebase/auth/FirebaseTokenVerifierImpl.java [hunk 18]
```diff
@@ -97,7 +99,7 @@ private FirebaseTokenVerifierImpl(Builder builder) {
   public FirebaseToken verifyToken(String token) throws FirebaseAuthException {
     IdToken idToken = parse(token);
     checkContents(idToken);
-    if (!isEmulatorMode()) {
+    if (!isEmulatorMode) {
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/FirebaseTokenVerifierImpl.java:102]: If you want you can make `isEmulatorMode` a local variable here instead of a class attribute.

**Review by hiranya911 — APPROVED** [review]
LGTM 👍

File: src/test/java/com/google/firebase/auth/EmulatorFirebaseTokenVerifierImplTest.java [hunk 19]
```diff
@@ -0,0 +1,328 @@
...
+    tokenVerifier.verifyToken(token);
+  }
+
+  @Test
+  public void testVerifyTokenIncorrectAlgorithm() {
+    String token = createTokenWithIncorrectAlgorithm();
+
+    try {
+      tokenVerifier.verifyToken(token);
+    } catch (FirebaseAuthException e) {
```
Comment (yuchenshi) [inline src/test/java/com/google/firebase/auth/EmulatorFirebaseTokenVerifierImplTest.java:86]: This shouldn't throw -- please update accordingly (similar to next test). (FYI it works right now because it doesn't `fail` when no error is thrown, but the intention is still wrong)

**Review by yuchenshi — APPROVED** [review]

File: src/test/java/com/google/firebase/auth/EmulatorFirebaseTokenVerifierImplTest.java (hunk 19 above)
Comment (ssbushi) [reply src/test/java/com/google/firebase/auth/EmulatorFirebaseTokenVerifierImplTest.java]: Good catch. Thanks!

File: src/main/java/com/google/firebase/auth/FirebaseTokenVerifierImpl.java (hunk 18 above)
Comment (ssbushi) [reply src/main/java/com/google/firebase/auth/FirebaseTokenVerifierImpl.java:101]: Done. Thanks!
